third_party/cached_httpfs/http_file_cache.cpp

```diff
@@ -0,0 +1,103 @@
+#include "http_file_cache.hpp"
+
+namespace duckdb {
+
+CachedFile::CachedFile(const string &cache_dir, FileSystem &fs, const std::string &key, bool cache_file) : fs(fs) {
+	file_name = cache_dir + "/" + key;
+
+	GetDirectoryCacheLock(cache_dir);
+
+	FileOpenFlags flags =
+	    FileFlags::FILE_FLAGS_READ | FileFlags::FILE_FLAGS_NULL_IF_NOT_EXISTS | FileLockType::READ_LOCK;
+	handle = fs.OpenFile(file_name, flags);
+	if (handle) {
+		initialized = true;
+		size = handle->GetFileSize();
+	} else if (cache_file) {
+		flags = FileFlags::FILE_FLAGS_WRITE | FileFlags::FILE_FLAGS_FILE_CREATE | FileLockType::WRITE_LOCK;
+		handle = fs.OpenFile(file_name, flags);
+	}
+
+	ReleaseDirectoryCacheLock();
+}
+
+CachedFile::~CachedFile() {
+	if (!initialized && handle) {
```

Comment:
The handle should probably be closed before we remove the file

Reply:
Also, I don't think this condition is correct, we use `initialized` to indicate that the file is ready for reading, not that it was created by this process.

Even when it was created by this process, that doesn't necessarily mean that the file can be removed as another process can still be using the file. I'm not entirely sure what the condition should be here, but perhaps we need a interprocess reference count for the file.

I guess if we try to remove the file while another process is still using it, do we hang until it's closed or is the file unlinked and scheduled for removal when the other process closes the file?

Reply:
We could be hitting open file descriptor limitations if we're caching any decent amount of files which is also something to keep in mind, but perhaps figuring that out is a future PR

Reply:
When file doesn't exists, i.e when we cache file it will open file with WRITE_LOCK so all other processes cannot open file until that lock is released. 
This logic is intended to clean up file if user interrupts caching (initialize = false & file handle != null) but as we, for now, don't have ability to do this i'll remove this logic in destructor.
As discussed, improvements for this extension will be done later, but it is good that we keep track of what could be potential problems.